src/garage/torch/policies/context_conditioned_policy.py

```diff
@@ -0,0 +1,268 @@
+        policy (garage.torch.policies.Policy): Policy used to train the
+            network.
+        use_ib (bool): True if latent context is not deterministic; false
+            otherwise
+        use_next_obs (bool): True if next observation is used in context
+            for distinguishing tasks; false otherwise.
+
+    """
+
+    def __init__(self, latent_dim, context_encoder, policy, use_ib,
+                 use_next_obs):
+        super().__init__()
+        self._latent_dim = latent_dim
+        self._context_encoder = context_encoder
+        self._policy = policy
+        self._use_ib = use_ib
+        self._use_next_obs = use_next_obs
+
+        # initialize buffers for z distribution and z
+        # use buffers so latent context can be saved along with model weights
+        # z_means and z_vars are the params for the gaussian distribution
+        # over latent task belief maintained in the policy; z is a sample from
+        # this distribution that the policy is conditioned on
+        self.register_buffer('z', torch.zeros(1, latent_dim))
+        self.register_buffer('z_means', torch.zeros(1, latent_dim))
+        self.register_buffer('z_vars', torch.zeros(1, latent_dim))
+
+        self.reset_belief()
+
+    def reset_belief(self, num_tasks=1):
+        """Reset q(z|c) to the prior and sample a new z from the prior.
+
+        Args:
+            num_tasks (int): Number of tasks.
+
+        """
+        # reset distribution over z to the prior
+        mu = torch.zeros(num_tasks, self._latent_dim).to(tu.global_device())
+        if self._use_ib:
+            var = torch.ones(num_tasks,
+                             self._latent_dim).to(tu.global_device())
+        else:
+            var = torch.zeros(num_tasks,
+                              self._latent_dim).to(tu.global_device())
+        self.z_means = mu
+        self.z_vars = var
+        # sample a new z from the prior
+        self.sample_from_belief()
+        # reset the context collected so far
+        self._context = None
+        # reset any hidden state in the encoder network (relevant for RNN)
+        self._context_encoder.reset(num_tasks)
+
+    def sample_from_belief(self):
+        """Sample z using distributions from current means and variances."""
+        if self._use_ib:
+            posteriors = [
+                torch.distributions.Normal(m, torch.sqrt(s)) for m, s in zip(
+                    torch.unbind(self.z_means), torch.unbind(self.z_vars))
+            ]
+            z = [d.rsample() for d in posteriors]
+            self.z = torch.stack(z)
+        else:
+            self.z = self.z_means
+
+    def detach_z(self):
+        """Disable backprop through z."""
+        self.z = self.z.detach()
+
+    def update_context(self, inputs):
+        """Append single transition to the current context.
+
+        Args:
+            inputs (dict): Dictionary of transition information in np arrays .
+
+        """
+        o, a, r, no, _, _ = inputs
```

Comment:
Is it possible for you to use https://github.com/rlworkgroup/garage/blob/master/src/garage/_dtypes.py#L410 for this? If not, can you accept a concrete set of args rather than an opaque dict?